### PR TITLE
Handle selectors correctly in VersionedParams().

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #15945

### CODE REVIEW

**Comment by wojtek-t** [comment]
@lavalamp - bumping to the top of you queue (this is blocking other PRs)

File: pkg/client/unversioned/request.go [hunk 1]
```diff
@@ -425,7 +425,36 @@ func (r *Request) VersionedParams(obj runtime.Object, convertor runtime.ObjectCo
 	}
 	for k, v := range params {
 		for _, vv := range v {
-			r.setParam(k, vv)
+			// TODO: Move it to setParam method, once we get rid of
+			// FieldSelectorParam & LabelSelectorParam methods.
+			value := vv
+			if k == unversioned.LabelSelectorQueryParam(r.apiVersion) && value == "" {
+				// Don't set an empty selector for backward compatibility.
```
Comment (lavalamp) [inline pkg/client/unversioned/request.go]: clarify the nature of the backwards compatibility?

File: pkg/client/unversioned/request.go [hunk 2]
```diff
@@ -425,7 +425,36 @@ func (r *Request) VersionedParams(obj runtime.Object, convertor runtime.ObjectCo
 	}
 	for k, v := range params {
 		for _, vv := range v {
-			r.setParam(k, vv)
+			// TODO: Move it to setParam method, once we get rid of
+			// FieldSelectorParam & LabelSelectorParam methods.
+			value := vv
```
Comment (lavalamp) [inline pkg/client/unversioned/request.go]: nit: just change vv -> value on line 427?

File: pkg/client/unversioned/request.go [hunk 3]
```diff
@@ -425,7 +425,36 @@ func (r *Request) VersionedParams(obj runtime.Object, convertor runtime.ObjectCo
...
+			if k == unversioned.FieldSelectorQueryParam(r.apiVersion) {
+				if value == "" {
+					// Don't set an empty selector for backward compatibility.
+					continue
+				}
+				// TODO: Filtering should be handled somewhere else.
+				selector, err := fields.ParseSelector(value)
+				if err != nil {
+					r.err = err
+					continue
```
Comment (lavalamp) [inline pkg/client/unversioned/request.go]: return r

File: pkg/client/unversioned/request.go [hunk 4]
```diff
@@ -425,7 +425,36 @@ func (r *Request) VersionedParams(obj runtime.Object, convertor runtime.ObjectCo
...
+				}
+				filteredSelector, err := selector.Transform(
+					func(field, value string) (newField, newValue string, err error) {
+						return fieldMappings.filterField(r.apiVersion, r.resource, field, value)
+					})
+				if err != nil {
+					r.err = err
+					continue
```
Comment (lavalamp) [inline pkg/client/unversioned/request.go]: return r

File: pkg/client/unversioned/request.go [hunk 5]
```diff
@@ -425,7 +425,36 @@ func (r *Request) VersionedParams(obj runtime.Object, convertor runtime.ObjectCo
...
+			}
+			if k == unversioned.FieldSelectorQueryParam(r.apiVersion) {
+				if value == "" {
+					// Don't set an empty selector for backward compatibility.
+					continue
+				}
+				// TODO: Filtering should be handled somewhere else.
+				selector, err := fields.ParseSelector(value)
+				if err != nil {
+					r.err = err
```
Comment (lavalamp) [inline pkg/client/unversioned/request.go]: It might be good to expand this, e.g. `r.err = fmt.Errorf("unparsable field selector: %v", err)`, so people can tell where the error came from more easily.

Same below.

Comment (wojtek-t) [inline pkg/client/unversioned/request.go]: Done.

File: pkg/client/unversioned/request.go (hunk 4 above)
Comment (wojtek-t) [inline pkg/client/unversioned/request.go]: Done.

File: pkg/client/unversioned/request.go (hunk 3 above)
Comment (wojtek-t) [inline pkg/client/unversioned/request.go]: Done.

File: pkg/client/unversioned/request.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/client/unversioned/request.go]: Done.

File: pkg/client/unversioned/request.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/client/unversioned/request.go]: We just didn't set the corresponding url param it that wasn't set. So I didn't want to explicitly set fieldSelector=&labelSelector= in every request.

**Comment by wojtek-t** [comment]
@lavalamp - PTAL

File: pkg/client/unversioned/request.go [hunk 6]
```diff
@@ -427,8 +427,36 @@ func (r *Request) VersionedParams(obj runtime.Object, convertor runtime.ObjectCo
 		return r
 	}
 	for k, v := range params {
-		for _, vv := range v {
-			r.setParam(k, vv)
+		for _, value := range v {
+			// TODO: Move it to setParam method, once we get rid of
+			// FieldSelectorParam & LabelSelectorParam methods.
+			if k == unversioned.LabelSelectorQueryParam(r.apiVersion) && value == "" {
+				// Don't set an empty selector for backward compatibility.
```
Comment (lavalamp) [inline pkg/client/unversioned/request.go]: I see. Perhaps we need to fix the converter or make something into a pointer so we can tell the difference between unspecified and deliberately set to ""? I don't *think* we make a distinction in query parameters, but we do for the label selectors in e.g. service objects-- unset means nothing and set-but-empty means everything. So this may be important to get right, or at least expand the comment?

Comment (wojtek-t) [inline pkg/client/unversioned/request.go]: The issue here is that in golang there is no way to differ "" and unspecified string.
I will expand this comment.

**Comment by wojtek-t** [comment]
@lavalamp - PTAL

File: pkg/client/unversioned/request.go (hunk 6 above)
Comment (lavalamp) [inline pkg/client/unversioned/request.go]: Right, and this may not be a big deal-- but you'd probably have to fix this a level up by making unversioned.ListOptions distinguish between unset selector and set-but-empty (by making it a pointer, if necessary).

Comment (wojtek-t) [inline pkg/client/unversioned/request.go]: Can we do it in a separate PR?
This PR is not introducing any regression.

Comment (lavalamp) [inline pkg/client/unversioned/request.go]: Sure, that's fine.

**Comment by lavalamp** [comment]
lgtm

**Comment by wojtek-t** [comment]
Just fixed the gofmt issue.

**Comment by wojtek-t** [comment]
@k8s-bot e2e test this please

**Comment by wojtek-t** [comment]
Tests passing, internal e2e are green.
Merging manually, because merge-bot has some internal problems:

    E1201 09:10:07.238529       1 github.go:1079] no mergeability information for "docs(getting-started): add AZ_LOCATION information to README" 17926, Skipping
    E1201 09:10:23.883793       1 github.go:63] Ran out of github API tokens. Sleeping for 38.70193685148333 minutes
    E1201 09:10:38.578980       1 github.go:63] Ran out of github API tokens. Sleeping for 38.45701710115 minutes

**Comment by wojtek-t** [comment]
I filed https://github.com/kubernetes/contrib/issues/300 for the merging issue.

File: pkg/client/unversioned/request.go (hunk 6 above)
Comment (smarterclayton) [inline pkg/client/unversioned/request.go]: This really needs to be in a converter.  I view this as an ugly side effect of unversioned.ListOptions (which I commented on the other that I don't believe is the right direction anymore).

Comment (wojtek-t) [inline pkg/client/unversioned/request.go]: Yes - but we can move it to converter, without getting back to versioned ListOptions, I think.

Comment (smarterclayton) [inline pkg/client/unversioned/request.go]: ListOPtions *should* be versioned is my argument.  Seeing this code, seeing
the field transformation code, it's basically versioning.  The API for that
object has to be stable, which is the whole point of what we had "internal"
objects before.  The rest of the path is wire transformation - saying that
there is no wire transformation for ListOptions is silly.  Creating a
different way of handling list options from all the other options is
basically creating an extra code path, not removing a code path.

Comment (wojtek-t) [inline pkg/client/unversioned/request.go]: I'm not saying that we shouldn't get back to versioned ListOptions.

What I'm saying is that we already had exactly the same code in Request::LabelSelector() and Request::FieldSelector before. So this code wasn't created - it was move from one place to the other.

That said - I agree it shouldn't be here. We should basically move "fieldmappig" to some more generic places (conversion-related) and once this is done, we can also move this code there.
So in my opinion, cleaning up this code is unrelated to getting back to ListOptions...

Comment (wojtek-t) [inline pkg/client/unversioned/request.go]: So in other words - we should definitely clean this up, even before getting back to versioned ListOptions - because those are orthogonal problems.

Comment (smarterclayton) [inline pkg/client/unversioned/request.go]: Yes they are orthogonal, but related in the sense of location.

There should be conversion functions for field selectors that know how to
target the field selectors.  The hard problem is that the type of the field
selector is a dynamic input, not a static input.  We may need to support
conversion options to Convert that define additional metadata that allows
the field selector conversion to target base on the input metadata
(essentially an object reference).  One scheme is not guaranteed to hold
all conversions either - so there's an inside out problem here we may have
to solve in the future.

Comment (wojtek-t) [inline pkg/client/unversioned/request.go]: I agree with the first part.

>
 One scheme is not guaranteed to hold
all conversions either - so there's an inside out problem here we may have
to solve in the future.

If that's the case, we will need to change this code too, because currently we only have one convertor here.

Sorry -  I probably don't see some consequence that you can see, so I'm trying to understand them (also note that I'm not opting for not changing it back to versioned - I'm just trying to understand what it would give us now...)
